rules/shared/package-json.js

```diff
@@ -0,0 +1,52 @@
+import fs from 'node:fs';
+import {findUpSync} from 'find-up-simple';
+
+const directoryCache = new Map();
+const dataCache = new Map();
+
+/**
+Finds the closest package.json file to the given directory and returns its path and contents.
+
+Caches the result for future lookups.
+
+@param dirname {string}
+@return {{ path: string, packageJson: Record<string, unknown> } | undefined}
+*/
+export function readPackageJson(dirname) {
+	if (directoryCache.has(dirname)) {
+		const packageJsonPath = directoryCache.get(dirname);
+
+		if (dataCache.has(packageJsonPath)) {
+			return {
+				path: packageJsonPath,
+				packageJson: dataCache.get(packageJsonPath),
+			};
+		}
+	}
+
+	const packageJsonPath = findUpSync('package.json', {cwd: dirname, type: 'file'});
+	if (!packageJsonPath) {
+		return;
```

Comment:
I think we should also cache the result if it's not found.

Same on line 45

Reply:
I added caching for missing file paths, but caching line 45 would mean that we would cache any errors in the file, which I believe would not be desired.

Especially when considering language servers

Reply:
Simplified little bit, [`5188d85` (#2619)](https://github.com/sindresorhus/eslint-plugin-unicorn/pull/2619/commits/5188d85427ca3e013416adbfb4949eb9c5d51004)